src/main/java/com/apple/spark/core/SparkPodNodeAffinityHelper.java

```diff
@@ -0,0 +1,221 @@
+/*
+ *
+ * This source file is part of the Batch Processing Gateway open source project
+ *
+ * Copyright 2022 Apple Inc. and the Batch Processing Gateway project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.apple.spark.core;
+
+import static com.apple.spark.core.Constants.*;
+
+import com.apple.spark.AppConfig;
+import com.apple.spark.api.SubmitApplicationRequest;
+import com.apple.spark.operator.*;
+import com.google.common.annotations.VisibleForTesting;
+import java.util.*;
+
+public class SparkPodNodeAffinityHelper {
+
+  public static final String K8S_ZONE_LABEL = "topology.kubernetes.io/zone";
+
+  public static NodeAffinity createNodeAffinityForSparkPods(
+      final SubmitApplicationRequest request,
+      final AppConfig appConfig,
+      final String queueName,
+      final AppConfig.SparkCluster sparkCluster,
+      final boolean isDriver) {
+
+    NodeAffinity nodeAffinity = new NodeAffinity();
+
+    NodeSelectorTerm nodeSelectorTermForRequired =
+        createNodeSelectorTermForRequired(request, appConfig, queueName, sparkCluster, isDriver);
+
+    if (nodeSelectorTermForRequired.getMatchExpressions() != null) {
+      nodeAffinity.setRequiredDuringSchedulingIgnoredDuringExecution(
+          new RequiredDuringSchedulingIgnoredDuringExecutionTerm(
+              new NodeSelectorTerm[] {nodeSelectorTermForRequired}));
+    }
```

Comment:
**Comment**: reserve room for `PreferredDuringSchedulingIgnoredDuringExecutionTerm`